internal/graph/dispatch_throttling_check_resolver.go

```diff
@@ -0,0 +1,123 @@
+package graph
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/pkg/telemetry"
+)
+
+// DispatchThrottlingCheckResolverConfig encapsulates configuration for dispatch throttling check resolver
+type DispatchThrottlingCheckResolverConfig struct {
+	TimerTickerFrequency time.Duration
+	Level                uint32
+	Rate                 uint32
+}
+
+// DispatchThrottlingCheckResolver will prioritize requests with fewer dispatches over
+// requests with more dispatches.
+// Initially, request's dispatches will not be throttled and will be processed
+// immediately. When the number of request dispatches is above the Level, the dispatches are placed
+// in the throttling queue. One item form the throttling queue will be processed every Nth ticker
+// (configured via config's Rate parameter).
+// This allows a check / list objects request to be gradually throttled.
+type DispatchThrottlingCheckResolver struct {
+	delegate        CheckResolver
+	config          DispatchThrottlingCheckResolverConfig
+	ticker          *time.Ticker
+	throttlingQueue chan bool
+	done            chan bool
+}
+
+var _ CheckResolver = (*DispatchThrottlingCheckResolver)(nil)
+
+var (
+	dispatchThrottlingResolverDelayMsHistogram = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "dispatch_throttling_resolver_delay_ms",
+		Help:                            "Time spent waiting for dispatch throttling resolver",
+		Buckets:                         []float64{1, 3, 5, 10, 25, 50, 100, 1000, 5000}, // Milliseconds. Upper bound is config.UpstreamTimeout.
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	}, []string{"grpc_service", "grpc_method"})
+)
+
+func NewDispatchThrottlingCheckResolver(
+	config DispatchThrottlingCheckResolverConfig) *DispatchThrottlingCheckResolver {
+	dispatchThrottlingCheckResolver := &DispatchThrottlingCheckResolver{
+		config:          config,
+		ticker:          time.NewTicker(config.TimerTickerFrequency),
+		throttlingQueue: make(chan bool),
```

Comment:
No need for a `bool` here and below for the `done` channel. We don't need a value, just a signal. So use the empty struct instead `struct{}`, because it has no overhead as compared to a bool type which takes a byte.

Reply:
updated